_posts/2025-11-19-cheri-or-cheriot.markdown

```diff
@@ -0,0 +1,128 @@
+
+These all provide the same core set of features.
+Most software isn't written for a single target instruction-set architecture (ISA), it's written assuming a set of language-level abstractions that can be represented in a particular target.
+Software written for CHERI in C (or any language that is higher-level than C) doesn't care the bit patterns of capabilities, how bounds or permissions are encoded, and so on.
+It cares only that you have some capability registers, that you can store capabilities in memory, and that the C language notion of a pointer can be represented by a CHERI capability for the target ISA.
+
+I like to use memory-management units (MMUs) as an analogy for CHERI.
+MIPS and SPARC had software-managed translation-lookaside buffers (TLBs).
+x86 and Arm had architectural page tables that hardware walked to fill the TLB automatically on miss.
+PowerPC had an MMU design that, in the words of Linus Torvalds, 'can be used to frighten small children'.
+All of these implemented the same set of abstractions in terms of processes, shared memory, copy-on-write regions, and so on.
+All of these are MMU designs, just as there are many possible CHERI implementations.
+
+There are lots of things like this in ISAs.
+Arm's Neon, PowerPC's AltiVec, and x86 SSE are all SIMD extensions, which provide similar programming models, but they all have quite different low-level details.
+It's quite possible to write code that targets all of them.
+
+CHERIoT is a concrete platform
+------------------------------
+
+CHERIoT is a complete ISA specification, co-designed with a software stack.
+In the same way that x86 page tables or MIPS' software-managed TLB are both concrete instances of the abstract idea of a memory-management unit, CHERIoT is a concrete instantiation of CHERI.
+
+This means that there is no 'CHERI or CHERIoT' choice.
+The real choices are between CHERI or not CHERI, or between CHERIoT and some other concrete instantiation of CHERI.
+Having worked on CHERI since 2012, I am obviously biased on the first of these: given a choice between CHERI and not-CHERI, you should definitely pick CHERI!
+The second choice (CHERIoT or some other CHERI variant) is more nuanced, as I'll discuss more in the rest of this post.
+There are good reasons for CHERIoT but there are also domains where it is not the best choice.
+
+CHERIoT is a *rich* CHERI platform
+----------------------------------
+
+A minimal CHERI specification provides pointer integrity, bounds enforcement, and some permissions.
+Pointer integrity means (among other things) that you can precisely differentiate between pointers and other kinds of non-pointer data in memory and so can build temporal safety on top.
+Most of the CHERI temporal safety work has provided users with protection against use-after-reallocation issues.
+A pointer in a CHERI system may point to deallocated memory, but it will not point to *reallocated* memory.
```

Comment:
```suggestion
A pointer in a CHERI system with Cornucopia-style heap temporal safety may point to `free()`-ed memory, but it will not point to *reallocated* memory.
```